datastore/threads/threadproxyds.nim

```diff
@@ -0,0 +1,338 @@
+    ctx[].res[].err(error)
+    return
+
+  ctx[].res[].ok(res)
+
+method has*(self: ThreadDatastore, key: Key): Future[?!bool] {.async.} =
+  var
+    signal = ThreadSignalPtr.new().valueOr:
+        return failure(error())
+
+    res = ThreadResult[bool]()
+    ctx = TaskCtx[bool](
+      ds: addr self.ds,
+      res: addr res,
+      semaphore: addr self.semaphore,
+      signal: signal)
+
+  proc runTask() =
+    self.tp.spawn hasTask(addr ctx, unsafeAddr key)
+
+  self.dispatchTask(ctx, runTask)
+  return success(res.get())
+
+proc delTask(ctx: ptr TaskCtx, key: ptr Key) =
+  defer:
+    discard ctx[].signal.fireSync()
+    ctx[].semaphore[].release()
+
+  ctx[].semaphore[].acquire()
+  without res =? (waitFor ctx[].ds[].delete(key[])).catch, error:
+    ctx[].res[].err(error)
+    return
+
+  ctx[].res[].ok()
+
+method delete*(
+  self: ThreadDatastore,
+  key: Key): Future[?!void] {.async.} =
+
+  var
+    signal = ThreadSignalPtr.new().valueOr:
+        return failure(error())
+
+    res = ThreadResult[void]()
+    ctx = TaskCtx[void](
+      ds: addr self.ds,
+      res: addr res,
+      semaphore: addr self.semaphore,
+      signal: signal)
+
+  proc runTask() =
+    self.tp.spawn delTask(addr ctx, unsafeAddr key)
+
+  self.dispatchTask(ctx, runTask)
+  return success()
+
+method delete*(
+  self: ThreadDatastore,
+  keys: seq[Key]): Future[?!void] {.async.} =
+
+  for key in keys:
+    if err =? (await self.delete(key)).errorOption:
+      return failure err
+
+  return success()
+
+proc putTask(
+  ctx: ptr TaskCtx,
+  key: ptr Key,
+  # data: DataBuffer,
+  data: ptr UncheckedArray[byte],
+  len: int) =
+  ## run put in a thread task
+  ##
+
+  defer:
+    discard ctx[].signal.fireSync()
+    ctx[].semaphore[].release()
+
+  ctx[].semaphore[].acquire()
+  without res =? (waitFor ctx[].ds[].put(key[], @(data.toOpenArray(0, len - 1)))).catch, error:
+    ctx[].res[].err(error)
+    return
+
+  ctx[].res[].ok()
+
+method put*(
+  self: ThreadDatastore,
+  key: Key,
+  data: seq[byte]): Future[?!void] {.async.} =
+
+  var
+    signal = ThreadSignalPtr.new().valueOr:
+        return failure(error())
+
+    res = ThreadResult[void]()
+    ctx = TaskCtx[void](
+      ds: addr self.ds,
+      res: addr res,
+      semaphore: addr self.semaphore,
+      signal: signal)
+
+  proc runTask() =
+    self.tp.spawn putTask(
+      addr ctx,
+      unsafeAddr key,
+      makeUncheckedArray(baseAddr data),
+      data.len)
+
+  self.dispatchTask(ctx, runTask)
+  return success()
+
+method put*(
+  self: ThreadDatastore,
+  batch: seq[BatchEntry]): Future[?!void] {.async.} =
+
+  for entry in batch:
+    if err =? (await self.put(entry.key, entry.data)).errorOption:
+      return failure err
+
+  return success()
+
+proc getTask(
+  ctx: ptr TaskCtx,
+  key: ptr Key) =
+  ## Run get in a thread task
+  ##
+
+  defer:
+    discard ctx[].signal.fireSync()
+    ctx[].semaphore[].release()
+
+  ctx[].semaphore[].acquire()
```

Comment:
Nice idea, this would keep the backend synchronous. 🤔 

Though it wouldn't provide back-pressure to the async side as async-requests could just keep queuing up no? Still need something to tie to the main chronos thread. 

On line 224, `signal = ThreadSignalPtr.new().valueOr` an errant async proc could still try to load say a 10,000 keys creating a new AsyncFD for each.

Reply:
Tho, the `acquire` should happen before the thread signal.